### PR TITLE
Add `OmegaConfLoader`

### PR DESCRIPTION
**Pull request by merelcht**
## Description
Added a new config loader class based on `AbstractConfigLoader` that uses `OmegaConf` to load and merge configuration. 

## Background 🔍
- https://github.com/kedro-org/kedro/issues/1657
- https://github.com/kedro-org/kedro/issues/1868

## Development notes 🏗
Some differences between the `OmegaConfLoader` and the other config loader classes:
- It's not recommended to use YAML anchors with `OmegaConf`, because they offer [variable interpolation](https://omegaconf.readthedocs.io/en/2.2_branch/usage.html#variable-interpolation). Hence I didn't copy over the test [`test_ignore_hidden_keys`](https://github.com/kedro-org/kedro/blob/main/tests/config/test_config.py#L179-L195)
- YAML parsing is done a bit differently. I added `ScannerError` with `ParserError` to catch this. 

Other notes:
- While writing a test for checking whether a file has the `.yml` or `.yaml` extension, I found out that `OmegaConf` can also handle `.json` files. So I've added `.json` as an acceptable extension for this loader as well. 
- After refactoring my initial implementation, I made sure not to use any of the helper methods from `config/common.py`. The methods in there are deeply nested and hard to follow. Usage of these methods makes it pretty much impossible for users to customise any configloader implementations. 

## Review notes ✍️
- I added changes incrementally, so you can review this PR by checking out each individual commit. 
- All commits starting from https://github.com/kedro-org/kedro/pull/2085/commits/1c12a87259aca2b72be7277c46116da7c0e0d3de are a refactor to not use any methods from `config/common.py`

### To Discuss 🗣
❓**Which built-in resolvers should be turned off?**

We decided to initially turn off all build-in resolvers. It's easier to introduce them step by step, but removing them would be a breaking change. In https://github.com/kedro-org/kedro/issues/1909 we will introduce environment variables for credentials. You can find a description of all built-in resolvers here: https://omegaconf.readthedocs.io/en/2.2_branch/custom_resolvers.html#built-in-resolvers
I feel like maybe it's okay to keep some resolvers turned on. For example `oc.dict` seems useful and not something we wouldn't want to add. I would like to hear what others think.

❓ **Is `OmegaConfLoader` the best name?**

I first wanted to call this config loader `YAMLConfigLoader` but then I found out that `OmegaConf` can also load `.json` files. Any better name suggestions?

❓ **Will `OmegaConfLoader` replace `ConfigLoader`?**

Follow up on the above question. I'm thinking that in `0.19.0` we should remove the `ConfigLoader` and `TemplatedConfigLoader` and rename this new `OmegaConfLoader` to `ConfigLoader`. If we do that, maybe the name of this new class doesn't matter so much for the `0.18` series.

## Checklist ✅

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [X] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/2085"></a>

### CODE REVIEW

File: kedro/config/omegaconf_config.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+                f"Invalid YAML file {config_filepath}, unable to read line {line}, "
+                f"position {cursor}."
+            ) from exc
+
+        _check_duplicate_keys(seen_file_to_keys, config_filepath, dict(single_config))
+        seen_file_to_keys[config_filepath] = single_config.keys()
+        aggregate_config.append(single_config)
+
+    if len(aggregate_config) > 1:
+        return dict(OmegaConf.merge(*aggregate_config))
```
Comment (noklam) [inline kedro/config/omegaconf_config.py:249]: Note for myself - need to revisit this later.

File: tests/config/test_omegaconf_config.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        assert params["param1"] == 1
+        assert catalog["trains"]["type"] == "MemoryDataSet"
+        assert catalog["cars"]["type"] == "pandas.CSVDataSet"
+        assert catalog["boats"]["type"] == "MemoryDataSet"
+        assert not catalog["cars"]["save_args"]["index"]
+
+    @use_config_dir
+    def test_load_local_config(self, tmp_path):
+        """Make sure that configs from `local/` override the ones
+        from `base/`"""
```
Comment (noklam) [inline tests/config/test_omegaconf_config.py:109]: Could you explain how this is testing `local` override from `base`? It almost looks the same as the previous test.

**Review by noklam — COMMENTED** [review]
Leave 1st round of comments - haven't fully understood the code but since this is a relatively large PR, it's nice to get some early comments.

File: kedro/config/__init__.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
     MissingConfigException,
 )
 from .config import ConfigLoader
+from .omegaconf_config import OmegaConfLoader
```
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/config/__init__.py:11]: You were too fast! 🏎️  I was still working on the description and have just added a comment about exactly this 😄

File: kedro/config/omegaconf_config.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            str(Path(self.conf_source) / run_env),
+        ]
+
+
+def _get_config_from_patterns_omegaconf(
+    conf_paths: Iterable[str],
+    patterns: Iterable[str] = None,
+) -> Dict[str, Any]:
+    """Recursively scan for configuration files, load and merge them using OmegaConf, and
+    return them in the form of a config dictionary.
```
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/config/omegaconf_config.py:152]: I'm not sure if a deprecation warning would make the most sense for that purpose? It's a new method so maybe a bit odd to immediately add a deprecation warning to it. I can just move this method into `config/common.py`.

File: kedro/config/omegaconf_config.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
+"""This module provides ``kedro.config`` with the functionality to load one
+or more configuration files of yaml or json type from specified paths through OmegaConf.
+"""
+import logging
+from pathlib import Path
+from typing import AbstractSet, Any, Dict, Iterable, List, Set  # noqa
+
+from omegaconf import OmegaConf
+from yaml.parser import ParserError
+from yaml.scanner import ScannerError
```
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/config/omegaconf_config.py:10]: They each catch slightly different syntax problems in `yaml` files. For example, this throws a `ScannerError`:
```yaml
bad:
config
```

And this throws a `ParserError`:
```yaml
example_iris_data:
              type: pandas.CSVDataSet
          filepath: data/01_raw/iris.csv
```

File: kedro/config/omegaconf_config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (noklam) [reply kedro/config/omegaconf_config.py:152]: You are right, I am thinking of something more lightweight. Something like this, or maybe just one sentence in the docstring.

https://www.attrs.org/en/21.1.0/backward-compatibility.html

[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/config/omegaconf_config.py:152]: That's a nice example! Let me explore some good options. I think in general we might want to clarify to users that any "private" methods aren't meant to be used directly.

File: kedro/config/omegaconf_config.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+            ) from exc
+
+        _check_duplicate_keys(seen_file_to_keys, config_filepath, dict(single_config))
+        seen_file_to_keys[config_filepath] = single_config.keys()
+        aggregate_config.append(single_config)
+
+    if len(aggregate_config) > 1:
+        return dict(OmegaConf.merge(*aggregate_config))
+
+    return config
```
Comment (noklam) [inline kedro/config/omegaconf_config.py:251]: This should be able to  handle case when len=1 ? Then the line 251 `return config` can be removed

Comment (noklam) [reply kedro/config/omegaconf_config.py:251]: More importantly, is `OmegaConf.merge` a syntatic suger of merging a list of dictionary? Is it possible to remove `aggregated_config` altogether and just merging the dictionary at each iteration?

Comment (merelcht) [reply kedro/config/omegaconf_config.py:251]: `OmegaConf.merge` throws an assertion error if the list is empty: https://github.com/omry/omegaconf/blob/master/omegaconf/omegaconf.py#L267. As far as I understand it's not just syntactic sugar and does the merging of configs in a considerate way.

Comment (merelcht) [reply kedro/config/omegaconf_config.py:251]: `OmegaConf.merge` throws an assertion error if the list is empty: https://github.com/omry/omegaconf/blob/master/omegaconf/omegaconf.py#L267. As far as I understand it's not just syntactic sugar and does the merging of configs in a considerate way.

File: kedro/config/omegaconf_config.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+        new_conf = _load_configs_omegaconf(config_filepaths=config_filepaths)
+
+        aggregate_config.append(new_conf)
+        processed_files |= set(config_filepaths)
+
+    if not processed_files:
+        raise MissingConfigException(
+            f"No files of YAML or JSON format found in {conf_paths} matching the glob "
+            f"pattern(s): {patterns}"
+        )
```
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/config/omegaconf_config.py:206]: Almost, but we use `processed_files` to check if stuff has been processed before so that we don't load it again and then we emit a log message: https://github.com/kedro-org/kedro/blob/main/kedro/config/common.py#L182-L187

File: tests/config/test_omegaconf_config.py (hunk 2 above)
Comment (merelcht) [reply tests/config/test_omegaconf_config.py]: I've changed this so the first two tests focus on verifying the `.get()` and `[]` access and the third test actually verifies that `local` config overrides `base`.

File: kedro/config/omegaconf_config.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+                f"No files of YAML or JSON format found in {base_path} or {env_path} matching"
+                f" the glob pattern(s): {[*self.config_patterns[key]]}"
+            )
+        return config
+
+    def __repr__(self):  # pragma: no cover
+        return (
+            f"OmegaConfLoader(conf_source={self.conf_source}, env={self.env}, "
+            f"config_patterns={self.config_patterns})"
+        )
```
Comment (SajidAlamQB) [inline kedro/config/omegaconf_config.py:174]: Slightly confused about what these strings are for.

File: kedro/config/omegaconf_config.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        except (ParserError, ScannerError) as exc:
+            assert exc.problem_mark is not None
+            line = exc.problem_mark.line
+            cursor = exc.problem_mark.column
+            raise ParserError(
+                f"Invalid YAML or JSON file {config_filepath}, unable to read line {line}, "
+                f"position {cursor}."
+            ) from exc
+
+        _check_duplicate_keys(seen_file_to_keys, config_filepath, dict(single_config))
```
Comment (SajidAlamQB) [inline kedro/config/omegaconf_config.py:227]: Do we still want to use this method from `common.py` if we will be reworking it?

**Review by SajidAlamQB — COMMENTED** [review]
This looks great thank you for this awesome work! Just had a few comments.

File: kedro/config/omegaconf_config.py (hunk 8 above)
Comment (merelcht) [reply kedro/config/omegaconf_config.py:174]: The `__repr__` method in Python is used to represent a Class object as a string (https://www.educative.io/answers/what-is-the-repr-method-in-python). It's useful for debugging and in the interactive ipython mode. 

I added the `__repr__` method for `ConfigLoader` and `TemplatedConfigLoader` in https://github.com/kedro-org/kedro/pull/1966 after @noklam pointed out they're not printed properly when using in interactive mode (see https://github.com/kedro-org/kedro/pull/1966#discussion_r1004477060). The one here in `OmegaConfLoader` is basically the same as for the other config loader classes.

File: kedro/config/omegaconf_config.py (hunk 9 above)
Comment (merelcht) [reply kedro/config/omegaconf_config.py:227]: Good question! The reason I used this method from `common.py` is because we do want this behaviour in the `OmegaConfLoader` and it's exactly the same as for the other config loaders. The other methods in `common.py` that I didn't end up using were very indirect and nested calls, so I refactored them for this config loader. In `0.19.0` I envision we'll get rid of `common.py` so we could move `_check_duplicate_keys` inside `omegaconf_config.py`. 

I could've refactored all methods in `common.py` but that felt like unnecessary work since ideally we wouldn't have this file full of helper methods. 

Do you think this makes sense or would you prefer there's no calls to `common.py` methods at all in this class?

Comment (SajidAlamQB) [reply kedro/config/omegaconf_config.py:227]: This makes sense, thank you. I agree we should move the `_check_duplicate_keys` method in `0.19.0` to `omegaconf_config.py`.

File: kedro/config/omegaconf_config.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+            _check_duplicate_keys(
+                seen_file_to_keys, config_filepath, dict(single_config)
+            )
+            seen_file_to_keys[config_filepath] = single_config.keys()
+            aggregate_config.append(single_config)
+
+        if len(aggregate_config) > 1:
+            return dict(OmegaConf.merge(*aggregate_config))
+
+        return config
```
Comment (noklam) [inline kedro/config/omegaconf_config.py:242]: Can we remove the special handling of `single_config` when `len(aggregate_config) == 1`? It should handle with len == 1 and thus no need to have an extra variable `config` and
```suggestion
        single_config = {}
        aggregate_config = []
        seen_file_to_keys = {}  # type: Dict[Path, AbstractSet[str]]

        for config_filepath in config_files_filtered:
            try:
                single_config = OmegaConf.load(config_filepath)
            except (ParserError, ScannerError) as exc:
                assert exc.problem_mark is not None
                line = exc.problem_mark.line
                cursor = exc.problem_mark.column
                raise ParserError(
                    f"Invalid YAML or JSON file {config_filepath}, unable to read line {line}, "
                    f"position {cursor}."
                ) from exc

            _check_duplicate_keys(
                seen_file_to_keys, config_filepath, dict(single_config)
            )
            seen_file_to_keys[config_filepath] = single_config.keys()
            aggregate_config.append(single_config)


        return dict(OmegaConf.merge(*aggregate_config))

```

File: tests/config/test_omegaconf_config.py [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+
+    @use_config_dir
+    def test_load_core_config_get_syntax(self, tmp_path):
+        """Make sure core config can be fetched with .get()"""
+        conf = OmegaConfLoader(str(tmp_path))
+        params = conf.get("parameters")
+        catalog = conf.get("catalog")
+
+        assert params["param1"] == 1
+        assert catalog["trains"]["type"] == "MemoryDataSet"
```
Comment (noklam) [inline tests/config/test_omegaconf_config.py:112]: 👍🏼

**Review by noklam — COMMENTED** [review]
It is cleaner now with some of the methods moving into the class. I just re-iterate on one of the comment, generally look great!

File: kedro/config/omegaconf_config.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+
+        Returns:
+            Dict[str, Any]:  A Python dictionary with the combined
+               configuration from all configuration files.
+        """
+
+        # Allow bypassing of loading config from patterns if a key and value have been set
+        # explicitly on the ``OmegaConfLoader`` instance.
+        if key in self:
+            return super().__getitem__(key)
```
Comment (AhdraMeraliQB) [inline kedro/config/omegaconf_config.py:135]: Why aren't we doing this check on the other config loader implementations?

Comment (merelcht) [reply kedro/config/omegaconf_config.py:135]: The only reason is that I just added it today and didn't think about the other implementations yet 😉 I'll add it.

**Review by noklam — APPROVED** [review]
Thank you for explaining! Nice work🌟

File: kedro/config/omegaconf_config.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        config = {}
+        aggregate_config = []
+        seen_file_to_keys = {}  # type: Dict[Path, AbstractSet[str]]
+
+        for config_filepath in config_files_filtered:
+            try:
+                single_config = OmegaConf.load(config_filepath)
+                config = single_config
+            except (ParserError, ScannerError) as exc:
+                assert exc.problem_mark is not None
```
Comment (idanov) [inline kedro/config/omegaconf_config.py:225]: What is this supposed to do? Could we get rid of the assert here?

File: kedro/config/omegaconf_config.py [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+                line = exc.problem_mark.line
+                cursor = exc.problem_mark.column
+                raise ParserError(
+                    f"Invalid YAML or JSON file {config_filepath}, unable to read line {line}, "
+                    f"position {cursor}."
+                ) from exc
+
+            _check_duplicate_keys(
+                seen_file_to_keys, config_filepath, dict(single_config)
+            )
```
Comment (idanov) [inline kedro/config/omegaconf_config.py:235]: Could we do this check only at the end of the loop, so we can give error message for all duplicates at once?

In fact, maybe we could keep only one variable, say `config_per_file`, mapping `filepath -> config`, and then check the duplicates in a separate loop, which will produce all errors at once.

After all `aggregate_config` is just `config_per_file.values()` and `seen_file_to_keys = {file: set(config.keys()) for file, config in config_per_file.items()}`. Updating two data structures in the same loop is harder to read.

File: kedro/config/omegaconf_config.py [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+        """
+
+        # Allow bypassing of loading config from patterns if a key and value have been set
+        # explicitly on the ``OmegaConfLoader`` instance.
+        if key in self:
+            return super().__getitem__(key)
+
+        # In the first iteration of the OmegaConfLoader we'll keep the resolver turned-off.
+        # It's easier to introduce them step by step, but removing them would be a breaking change.
+        self._clear_omegaconf_resolvers()
```
Comment (idanov) [inline kedro/config/omegaconf_config.py:139]: I wonder whether we could call this on `OmegaConfLoader.__init__` once, instead on each request for config. WDYT the implications of this could be?

File: kedro/config/omegaconf_config.py [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+
+    """
+
+    def __init__(
+        self,
+        conf_source: str,
+        env: str = None,
+        runtime_params: Dict[str, Any] = None,
+        config_patterns: Dict[str, List[str]] = None,
+        *,
```
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/config/omegaconf_config.py:78]: No they're not, but we did add it to `ConfigLoader` and `TemplatedConfigLoader`. Do you think it shouldn't be an argument at all in `__init__`?

File: kedro/config/omegaconf_config.py (hunk 13 above)
Comment (merelcht) [reply kedro/config/omegaconf_config.py:225]: As far as I understand this is to verify whether we can get `exc.problem_mark.line` and `exc.problem_mark.column` for nicely formatting the error message. Perhaps the assert isn't necessary here.

File: kedro/config/omegaconf_config.py (hunk 14 above)
Comment (merelcht) [reply kedro/config/omegaconf_config.py:235]: So instead you would load all the files and then compare all of them against each other to find duplicates? Wouldn't that me a lot more inefficient? I guess the benefit is that you would immediately know all duplicates, but the downside is that if you have a lot of config everything needs to be loaded first and then checked before you know something is wrong.

File: kedro/config/omegaconf_config.py (hunk 15 above)
Comment (merelcht) [reply kedro/config/omegaconf_config.py:139]: Yes actually I think that makes more sense. Otherwise you could get a situation where if the `__getitem__` function has been called users could technically use the resolvers. Seems unlikely, but I think it's better to move it up to `__init__`.

File: kedro/config/omegaconf_config.py (hunk 13 above)
Comment (merelcht) [reply kedro/config/omegaconf_config.py:225]: Ahh I think this might have been put there for the `mypy` linter:

```
kedro/config/omegaconf_config.py:224: error: Item "None" of "Optional[Mark]" has no attribute "line"
kedro/config/omegaconf_config.py:225: error: Item "None" of "Optional[Mark]" has no attribute "column"
Found 2 errors in 1 file (checked 265 source files)
```

I'll just add an ignore.

File: kedro/config/omegaconf_config.py (hunk 14 above)
Comment (merelcht) [reply kedro/config/omegaconf_config.py]: I've rewritten this to now check duplicates between all config files in a directory + subdirectories.

**Review by idanov — APPROVED** [review]
